server/image_pull.go

```diff
@@ -27,38 +27,75 @@ func (s *Server) PullImage(ctx context.Context, req *pb.PullImageRequest) (resp
 		image = img.Image
 	}
 
-	sourceCtx := *s.config.SystemContext // A shallow copy we can modify
+	pullArgs := pullArguments{image: image}
 	if req.GetAuth() != nil {
 		username := req.GetAuth().Username
 		password := req.GetAuth().Password
 		if req.GetAuth().Auth != "" {
 			username, password, err = decodeDockerAuth(req.GetAuth().Auth)
 			if err != nil {
-				log.Debugf(ctx, "error decoding authentication for image %s: %v", img, err)
+				log.Debugf(ctx, "error decoding authentication for image %s: %v", image, err)
 				return nil, err
 			}
 		}
 		// Specifying a username indicates the user intends to send authentication to the registry.
 		if username != "" {
-			sourceCtx.DockerAuthConfig = &types.DockerAuthConfig{
+			pullArgs.credentials = types.DockerAuthConfig{
 				Username: username,
 				Password: password,
 			}
 		}
 	}
 
+	// We use the server's pullCache to record which images are currently being
+	// pulled. This allows for avoiding pulling the same image in parallel.
+	// Hence, if a given image is currently being pulled, we queue into the
+	// pullOperation's waitgroup and wait for the pulling goroutine to unblock
+	// us and re-use its results.
+	pullOp, pullInProcess := s.pullOperation(&pullArgs)
+	if !pullInProcess {
+		go func() {
+			pullOp.imageRef, pullOp.err = s.pullImage(ctx, &pullArgs)
+			// Unblock waiting goroutines and delete the entry from the cache.
+			s.pullOperationsLock.Lock()
+			defer s.pullOperationsLock.Unlock()
+			defer pullOp.wg.Done()
+			delete(s.pullOperationsInProgress, pullArgs)
```

Comment:
Hum… if _either_ of the `.Done()` or the `delete` crashed, some waiter could block indefinitely. I can’t see how we can win here, we might just as well let the `panic` (at least in either of the two) bubble up and let Systemd restart us (if that’s what happens).

Reply:
Actually, this is already the best it can get: `wg.Done()` ensures the caller doesn’t block, and leaving the operation in the map will let any future callers to find it, not block, and progress; they’ll ”only” be unable to try the same pull again, which is pretty good considering the failure.